syntax = "proto3";

package types;

// Общие перечисления
enum Role {
    PASSENGER = 0;
    CAPTAIN = 1;
    DOCTOR = 2;
    GUNMEN = 3;
    ENGINEER = 4;
    SCIENTIST = 5;
    SIGNALMEN = 6;
	ASSISTANT = 101;
	DEP_DOCTOR = 102;
	DEP_GUNMEN = 103;
	DEP_ENGINEER = 104;
	DEP_SCIENTIST = 105;
	DEP_SIGNALMEN = 106;
}

enum Action {
	USE_ITEM = 0;
	USE_ULT = 1;
	USE_BADGE = 2;
	ATTACK = 3;
	INFECT = 4;
	VOTE = 5;
	UNVOTE = 6;
	WAIT = 7;
	UP = 8;
	DOWN = 9;
}

enum Item {
	BADGE = 0;
	ROTATION = 1;
	BLASTER = 2;
	INJECTOR = 3;
	NOTEBOOK = 4;
	BATTERY = 5;
	SCANNER = 6;
	MOP = 7;
	FETUS = 8;
}

enum PlayerStatus {
	PS_UP = 0;
	PS_DOWN = 1;
	PS_DEAD = 2;
}

enum EventType {
  ET_USE_ITEM = 0;
  ET_USE_ULT = 1;
  ET_USE_BADGE = 2;
  ET_GET_ROLE = 3;
  ET_DEL_ROLE = 4;
  ET_INGAME_MESSAGE = 5;
  ET_GET_DUTY = 6;
  ET_CONNECT = 7;
  ET_DISCONNECT = 8;
  ET_DIED = 9;
  ET_DOWN = 10;
  ET_UP = 11;
  ET_VOTE = 12;
  ET_UNVOTE = 13;
}

enum VotingType {
	FOR_ALIEN = 0;
	FOR_CAPTAIN = 1;
	FOR_DOCTOR = 2;
	FOR_GUNMEN = 3;
	FOR_ENGINEER = 4;
	FOR_SCIENTIST = 5;
	FOR_SIGNALMEN = 6;
	
}

enum TargetType {
	TAT_NAME = 0;
	TAT_ITEM = 1;
	TAT_ROLE = 2;
}

// Метасообщения для таргетов
message ITarget {
	TargetType type = 1;
	string name = 2;
	Item item = 3;
	Role role = 4;
}

// Общие сообщения
message Event {
	EventType event = 1;
	ITarget event_who = 2;
	ITarget event_object = 3;
	repeated ITarget event_target = 4;
	string payload = 5;
}